### ISSUE TITLE
Wide character issue introduced between 5.3.3.6 and 5.3.3.9

### ISSUE THREAD

**Issue by jonandr** [desjarlais/Scintilla.NET#111]
**Describe the bug**
Hi, first of all, thanks for taking up the job of maintaining Scintilla.NET.
There is some kind of issue in the way which the package deals with wide characters. It affects searching and indicator fills, and probably much more.
I am not well versed in the internals of Scintilla so I will demonstrate with an example.

**To Reproduce**
Create a new project.
Add a Scintilla control and a button.
In Form1_Load, add the following code:
```
scintilla1.Text = "ä ä ä";
```
Note that ä is encoded as 0xC3 0xA4 in UTF8. I doubt the exact characters matter, the string just has to be of the form "{wide} {wide} {wide}".

In the button's click handler, add the following code:

```
scintilla1.IndicatorCurrent = (int)8;
scintilla1.IndicatorClearRange(0, scintilla1.TextLength);
scintilla1.IndicatorFillRange(0, 1);
scintilla1.IndicatorFillRange(2, 1);
scintilla1.IndicatorFillRange(4, 1);
```

The intention here is to underline every "ä" when the button is clicked.

**Expected behavior**
Every ä should have line under it, like so:

<img width="402" alt="5 3 3 6" src="https://github.com/desjarlais/Scintilla.NET/assets/22964455/1ccd4339-c6f8-429a-b4ad-a8a061807fbf">

This is what happens in version 5.3.3.6.

However, starting with version 5.3.3.9, it actually looks like this:

<img width="402" alt="5 3 3 16" src="https://github.com/desjarlais/Scintilla.NET/assets/22964455/582f14f2-da49-4922-8786-5613167da6fa">

It looks the same in 5.3.3.16. It looks like it's just underlining the first ä and the space. It seems very likely that the underlying issue is due to the handling of wide characters.

**Desktop (please complete the following information):**
 - OS: Windows
 - Version 11

**Additional context**
Possibly related to #65 ?

**Comment by desjarlais** [comment]
Thank you for the comment about taking over as maintainer.  I'll do my best to keep it going.

As for this issue, so far, I'm not reproducing, I'm pretty sure I'm following it the way you describe...am I missing something?

![image](https://github.com/desjarlais/Scintilla.NET/assets/9721728/6f4597a7-d822-40f4-ab45-47715f8856e1)

**Comment by jonandr** [comment]
Hmm, yes, the code is correct.

Maybe it's something related to the solution settings?

Does it reproduce if you try this solution:

https://github.com/jonandr/scintillanet-111

I'm using VS 2022.

**Comment by desjarlais** [comment]
I tried with your solution and even brought in the nuget for 5.3.3.6, but I don't see any text get underlined.  

What are you expecting to apply the underline to the character in your sample?

**Comment by jonandr** [comment]
Sorry, I put the part which performs the underlining on a button click, that may have been confusing.
I have pushed a new version of the solution now, can you try it?

The new version runs the underlining immediately on Form1_Load instead.
So hopefully it will show the issue just by running it.

So to clarify:
When I run the latest version of the repo (which uses 5.3.3.16 by default), then I got the following:

![bild](https://github.com/desjarlais/Scintilla.NET/assets/22964455/58d9d468-4fbf-40ae-a1b2-918d35b73dea)

If I change the nuget to 5.3.36, then I get this:

![bild](https://github.com/desjarlais/Scintilla.NET/assets/22964455/f8914cf4-7e19-41a5-9f8f-bab933352b49)

Do you get the same?

**Comment by desjarlais** [comment]
I launch the latest version and I see this:

![image](https://github.com/desjarlais/Scintilla.NET/assets/9721728/067ddad5-bb1f-4869-9c54-528bcd1fc21e)

**Comment by jonandr** [comment]
Interesting - I was trying to minimize the test case, and since I got the underlines on my machine I just figured that that was the Scintilla default. But it must be machine dependent.

I've changed the solution again - can you try it one last time? This time the setup should be exactly like the wiki says:

https://github.com/desjarlais/Scintilla.NET/wiki/Find-and-Highlight-Words

On my machine, the problem is still there

5.3.3.6:

![bild](https://github.com/desjarlais/Scintilla.NET/assets/22964455/61dbcb96-12d4-4b2a-a95e-f6f75c5c1d1c)

5.3.3.16:

![bild](https://github.com/desjarlais/Scintilla.NET/assets/22964455/60749034-f7d3-40d9-87b4-f396d0b4d458)

**Comment by desjarlais** [comment]
I do see that with your latest example.  So you want the same behavior as 5.3.3.6?

**Comment by jonandr** [comment]
Yes, certainly. Speaking purely as a library user I would not expect there to be a huge, breaking change happening between 5.3.3.6 and 5.3.3.16 :)

To be clear, I suspect something has gone very seriously wrong with the wide character handling between these versions. It's not just that indicators have quietly changed. For instance, add this snippet at the end of Form1_Load:

```
            scintilla1.TargetStart = 0;
            scintilla1.TargetEnd = scintilla1.TextLength;
            scintilla1.SearchFlags = SearchFlags.None;
            while (scintilla1.SearchInTarget("ä") != -1)
            {
                scintilla1.TargetStart = scintilla1.TargetEnd;
                scintilla1.TargetEnd = scintilla1.TextLength;
            }
```

In 5.3.3.6, it just finds all three occurrences of "ä", as expected.
But in 5.3.3.16 it goes into an *infinite loop*.

So you literally can't use Scintilla's search functionality to search for wide character strings anymore. I mean, whatever change has been made is tantamount to dropping support for wide characters.

**Comment by desjarlais** [comment]
It was the change I made to the CharToBytePosition function.  I made that for 65 to handle emojis, but I think I need to move that out into a separate function.  CharToBytePosition has about 63 references and I had concerns early on when looking at fixing 65 that something like this could happen.  I think I'll move the fix for 65 into a new function and leave CharToBytePosition in its pre 5.3.3.6 state which should bring back all previous behavior.

**Comment by desjarlais** [comment]
here is a screenshot of your sample with a soon to be released version 5.3.3.17 including the changes I mentioned above:

![image](https://github.com/desjarlais/Scintilla.NET/assets/9721728/daa1fe14-39c5-45c8-b7f5-ed25ef809cc0)

This looks good?

**Comment by jonandr** [comment]
Looks good!

**Comment by desjarlais** [comment]
This should be fixed in 5.3.3.17, thank you for reporting the issue and the repro.
